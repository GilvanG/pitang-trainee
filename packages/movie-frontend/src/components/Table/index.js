import { Table as TableMantine, Button } from "@mantine/core";

const Table = ({ actions = [], columns = [], rows = [] }) => {
  const hasActions = Boolean(actions.length);

  return (
    <TableMantine highlightOnHover>
      <thead>
        <tr>
          {columns.map((column, index) => (
            <th key={index}>{column.value}</th>
          ))}

          {hasActions && <th>Actions</th>}
        </tr>
      </thead>
      <tbody>
        {rows.map((row, index) => (
          <tr key={index}>
            {columns.map((column, index) => {
              const value = row[column.key];

              return (
                <td key={index}>
                  {typeof column.render === "function"
                    ? column.render(value)
                    : value}
                </td>
              );
            })}

            {hasActions && (
              <td>
                {actions.map((action, index) => (
                  <Button
                    mr={8}
                    key={index}
                    leftIcon={action.icon}
                    onClick={() => action.onClick(row)}
                    variant="white"
                    color={action.color}
                  >
                    {action.name}
                  </Button>
                ))}
              </td>
            )}
          </tr>
        ))}
      </tbody>
    </TableMantine>
  );
};

export default Table;
