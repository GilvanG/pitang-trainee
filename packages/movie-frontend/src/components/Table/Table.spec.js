import { cleanup, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Table from ".";

const columns = [
  {
    key: "id",
    name: "ID",
  },
  {
    key: "name",
    name: "Name",
  },
];

const rows = [
  {
    id: "f18c9261-cc18-4d89-aefa-be1bfee1dca8",
    name: "Product One",
  },
  {
    id: "5f7baefe-4139-4bf1-b53a-a6432a4ec781",
    name: "Product Two",
  },
];

describe("Table", () => {
  beforeEach(() => {
    cleanup();
  });

  it("Render Table", () => {
    const { asFragment, container } = render(<Table />);

    expect(asFragment()).toMatchSnapshot();

    const thead = container.querySelector("thead");
    const tHeadRows = thead.querySelectorAll("tr");

    const tbody = container.querySelector("tbody");
    const tBodyRows = tbody.querySelectorAll("tr");

    expect(tHeadRows.length).toBe(1);
    expect(tBodyRows.length).toBe(0);
  });

  it("Render Table with Columns and Rows", () => {
    const { container, queryByText } = render(
      <Table columns={columns} rows={rows} />
    );

    const thead = container.querySelector("thead");
    const tHeadRows = thead.querySelectorAll("tr");

    const tbody = container.querySelector("tbody");
    const tBodyRows = tbody.querySelectorAll("tr");

    expect(tHeadRows.length).toBe(1);
    expect(tBodyRows.length).toBe(2);

    expect(queryByText("Product One")).toBeTruthy();
    expect(queryByText("Product Two")).toBeTruthy();
    expect(queryByText("Product Three")).toBeFalsy();
  });

  it("Render Table with Columns and Rows and Actions", () => {
    const editFn = jest.fn();

    const actions = [
      {
        color: "white",
        name: "Edit",
        onClick: editFn,
      },
    ];

    const { getAllByText } = render(
      <Table actions={actions} columns={columns} rows={rows} />
    );

    const [editProduct1, editProduct2] = getAllByText("Edit");

    expect(editFn).not.toHaveBeenCalled();

    userEvent.click(editProduct1);

    expect(editFn).lastCalledWith({
      id: "f18c9261-cc18-4d89-aefa-be1bfee1dca8",
      name: "Product One",
    });

    expect(editFn).toHaveBeenCalledTimes(1);

    userEvent.click(editProduct2);

    expect(editFn).lastCalledWith({
      id: "5f7baefe-4139-4bf1-b53a-a6432a4ec781",
      name: "Product Two",
    });

    expect(editFn).toHaveBeenCalledTimes(2);
  });
});
