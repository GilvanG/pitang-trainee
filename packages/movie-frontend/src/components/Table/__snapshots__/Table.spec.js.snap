// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Render Table 1`] = `
<DocumentFragment>
  <table
    class="mantine-Table-root __mantine-ref-hover mantine-Table-hover mantine-clz1g3"
  >
    <thead>
      <tr />
    </thead>
    <tbody />
  </table>
</DocumentFragment>
`;
